File: Distribution-Grid-EV-CA/R_Yanning/02_03_sample_EV hh.R
library(data.table)

TAZ.hh.sd <- fread('data/mobility data/CSTDM processed/SDPTM_hh_home_TAZ.csv')
TAZ.hh.ld <- fread('data/mobility data/CSTDM processed/LDPTM_hh_home_TAZ.csv')
TAZ.hh.ld.accegr <- fread('data/mobility data/CSTDM processed/LDPTM_AccEgr_hh_home_TAZ.csv')
TAZ.hh.etm <- fread('data/mobility data/CSTDM processed/ETM_hh_home_TAZ.csv')

TAZ.hh.share <- fread('data/mobility data/EV Toolbox/evhh_share_TAZ.csv')

CA.hh.share <- TAZ.hh.share[,.(EVhh_CA=sum(EVhh_TAZ),hh_CA=sum(hh_TAZ)),by=.(year)]
CA.hh.share[,EVhh_share:=EVhh_CA/hh_CA]

# length(intersect(unique(TAZ.hh.share$TAZ),unique(TAZ.hh.sd$HomeZone)))
# length(unique(TAZ.hh.sd$HomeZone))
# length(unique(TAZ.hh.share$TAZ))

########### EV hh count in each TAZ #######################################
# short distance
TAZ.hh.sd.count <- TAZ.hh.sd[,.(n_hh=.N),by=.(HomeZone)]
TAZ.EVhh.sd.count <- data.table()
for (y in 2022:2045){
  TAZ.EVhh.sd.count <- rbind(TAZ.EVhh.sd.count,TAZ.hh.sd.count[,year:=y])
}
TAZ.EVhh.sd.count <- merge(x=TAZ.EVhh.sd.count,y=TAZ.hh.share[,c("year","TAZ","EVhh_share")],
                           by.x=c("year","HomeZone"),by.y=c("year","TAZ"))
TAZ.EVhh.sd.count[,n_EVhh:=n_hh*EVhh_share] 
TAZ.EVhh.sd.count[,n_EVhh_newf:=n_EVhh-shift(n_EVhh),by=HomeZone] 
TAZ.EVhh.sd.count[year==2022,n_EVhh_newf:=n_EVhh]
TAZ.EVhh.sd.count[,n_EVhh_new:=floor(n_EVhh_newf)]
# long distance
TAZ.hh.ld.count <- TAZ.hh.ld[,.(n_hh=.N),by=.(HomeZone)]
TAZ.EVhh.ld.count <- data.table()
for (y in 2022:2045){
  TAZ.EVhh.ld.count <- rbind(TAZ.EVhh.ld.count,TAZ.hh.ld.count[,year:=y])
}
TAZ.EVhh.ld.count <- merge(x=TAZ.EVhh.ld.count,y=TAZ.hh.share[,c("year","TAZ","EVhh_share")],
                           by.x=c("year","HomeZone"),by.y=c("year","TAZ"))
TAZ.EVhh.ld.count[,n_EVhh:=n_hh*EVhh_share] 
TAZ.EVhh.ld.count[,n_EVhh_newf:=n_EVhh-shift(n_EVhh),by=HomeZone] 
TAZ.EVhh.ld.count[year==2022,n_EVhh_newf:=n_EVhh]
TAZ.EVhh.ld.count[,n_EVhh_new:=floor(n_EVhh_newf)]
# long distance acc/egr
TAZ.hh.ld.accegr.count <- TAZ.hh.ld.accegr[,.(n_hh=.N),by=.(HomeZone)]
TAZ.EVhh.ld.accegr.count <- data.table()
for (y in 2022:2045){
  TAZ.EVhh.ld.accegr.count <- rbind(TAZ.EVhh.ld.accegr.count,TAZ.hh.ld.accegr.count[,year:=y])
}
TAZ.EVhh.ld.accegr.count <- merge(x=TAZ.EVhh.ld.accegr.count,y=TAZ.hh.share[,c("year","TAZ","EVhh_share")],
                           by.x=c("year","HomeZone"),by.y=c("year","TAZ"))
TAZ.EVhh.ld.accegr.count[,n_EVhh:=n_hh*EVhh_share] 
TAZ.EVhh.ld.accegr.count[,n_EVhh_newf:=n_EVhh-shift(n_EVhh),by=HomeZone] 
TAZ.EVhh.ld.accegr.count[year==2022,n_EVhh_newf:=n_EVhh]
TAZ.EVhh.ld.accegr.count[,n_EVhh_new:=floor(n_EVhh_newf)]

# ext - keep the zones (freeways), use aggregate EV hh share in CA (aggressive)
#TAZ.hh.etm.count <- TAZ.hh.etm[,.(n_hh=.N)]
TAZ.hh.etm.count <- TAZ.hh.etm[,.(n_hh=.N),by=.(HomeZone)]
TAZ.EVhh.etm.count <- data.table()
for (y in 2022:2045){
  TAZ.EVhh.etm.count <- rbind(TAZ.EVhh.etm.count,TAZ.hh.etm.count[,year:=y])
}
TAZ.EVhh.etm.count <- merge(x=TAZ.EVhh.etm.count,y=CA.hh.share[,c("year","EVhh_share")],
                           by="year")
TAZ.EVhh.etm.count[,n_EVhh:=n_hh*EVhh_share] 
TAZ.EVhh.etm.count[,n_EVhh_newf:=n_EVhh-shift(n_EVhh),by=HomeZone] 
TAZ.EVhh.etm.count[year==2022,n_EVhh_newf:=n_EVhh]
TAZ.EVhh.etm.count[,n_EVhh_new:=floor(n_EVhh_newf)]

######### sample EV hh in each TAZ each year ####################################
# NOTE: the hhs are NEW EVhh in each year (need to accumulate when used later)

sample.EVhh <- function(TAZ.hh,TAZ.EVhh.count){
  TAZ.EVhh <- data.table()
  TAZ.hh.temp <- TAZ.hh[HomeZone%in%TAZ.EVhh.count$HomeZone]
  set.seed(42)
  for (y in 2022:2045){
    TAZ.EVhh.pool <- merge(x=TAZ.hh.temp,y=TAZ.EVhh.count[year==y,c("year","HomeZone","n_EVhh_new")],
                              by="HomeZone")
    
    TAZ.EVhh.temp <- TAZ.EVhh.pool[,sample(x=SerialNo,size=min(n_EVhh_new)),by=HomeZone]
    TAZ.EVhh.temp[,year:=y]
    setnames(TAZ.EVhh.temp,"V1","hhID")
    
    TAZ.EVhh <- rbind(TAZ.EVhh,TAZ.EVhh.temp)
    
    TAZ.hh.temp <- TAZ.hh.temp[!(SerialNo%in%TAZ.EVhh$hhID)] 
  }
  return(TAZ.EVhh)
}

# long distance
TAZ.EVhh.ld <- sample.EVhh(TAZ.hh.ld,TAZ.EVhh.ld.count)
fwrite(TAZ.EVhh.ld,file='data/mobility data/CSTDM processed/EVhh_new_LDPTM_sample42.csv',row.names=FALSE)

# short distance
TAZ.EVhh.sd <- sample.EVhh(TAZ.hh.sd,TAZ.EVhh.sd.count)
fwrite(TAZ.EVhh.sd,file='data/mobility data/CSTDM processed/EVhh_new_SDPTM_sample42.csv',row.names=FALSE)

# long distance access/egress
TAZ.EVhh.ld.accegr <- sample.EVhh(TAZ.hh.ld.accegr,TAZ.EVhh.ld.accegr.count)
fwrite(TAZ.EVhh.ld.accegr,file='data/mobility data/CSTDM processed/EVhh_new_LDPTM_AccEgr_sample42.csv',row.names=FALSE)

# etm 
TAZ.EVhh.etm <- sample.EVhh(TAZ.hh.etm,TAZ.EVhh.etm.count)
fwrite(TAZ.EVhh.etm,file='data/mobility data/CSTDM processed/EVhh_new_ETM_sample42.csv',row.names=FALSE)

File: Distribution-Grid-EV-CA/02_03_sample_EV_hh.py
"""
02_03_sample_EV_hh.py — Python port of ``02_03_sample_EV hh.R``.

Samples new EV households per TAZ per year for each trip source.
"""

from __future__ import annotations

import os

import numpy as np
import pandas as pd

IN_DIR = "data/mobility_data/CSTDM_processed"
SHARE_CSV = "data/mobility_data/EV_Toolbox/evhh_share_TAZ.csv"
YEARS = list(range(2020, 2045))
SEED = 42


def _expand_years(count_tbl: pd.DataFrame) -> pd.DataFrame:
    return pd.concat(
        [count_tbl.assign(year=y) for y in YEARS], ignore_index=True
    )


def _calc_new_evhh(count_tbl: pd.DataFrame, share_tbl: pd.DataFrame, join_cols) -> pd.DataFrame:
    out = pd.merge(count_tbl, share_tbl, on=join_cols)
    out["n_EVhh"] = out["n_hh"] * out["EVhh_share"]
    out = out.sort_values(["HomeZone", "year"])
    out["n_EVhh_newf"] = out.groupby("HomeZone")["n_EVhh"].diff()
    out.loc[out["year"] == YEARS[0], "n_EVhh_newf"] = out["n_EVhh"]
    out["n_EVhh_new"] = np.floor(out["n_EVhh_newf"]).astype("Int64")
    return out


def sample_evhh(taz_hh: pd.DataFrame, taz_evhh_count: pd.DataFrame) -> pd.DataFrame:
    """Port of R ``sample.EVhh`` — samples hhIDs without replacement within a
    HomeZone, cumulatively across years (drawn hhIDs are removed from the pool).
    """
    pool = taz_hh[taz_hh["HomeZone"].isin(taz_evhh_count["HomeZone"].unique())].copy()
    rng = np.random.default_rng(SEED)
    drawn_ids: set = set()
    rows: list[pd.DataFrame] = []
    for y in YEARS:
        ycount = taz_evhh_count.loc[
            taz_evhh_count["year"] == y, ["HomeZone", "n_EVhh_new"]
        ]
        merged = pd.merge(pool, ycount, on="HomeZone")
        sampled_parts = []
        for zone, g in merged.groupby("HomeZone", sort=False):
            n = int(g["n_EVhh_new"].min())
            n = max(0, min(n, len(g)))
            if n == 0:
                continue
            idx = rng.choice(len(g), size=n, replace=False)
            pick = g.iloc[idx][["SerialNo"]].rename(columns={"SerialNo": "hhID"})
            pick["HomeZone"] = zone
            sampled_parts.append(pick)
        if sampled_parts:
            ydraw = pd.concat(sampled_parts, ignore_index=True)
            ydraw["year"] = y
            rows.append(ydraw)
            drawn_ids.update(ydraw["hhID"].tolist())
            pool = pool.loc[~pool["SerialNo"].isin(drawn_ids)].copy()
    return pd.concat(rows, ignore_index=True) if rows else pd.DataFrame(columns=["HomeZone", "hhID", "year"])


def main() -> None:
    sd = pd.read_csv(os.path.join(IN_DIR, "SDPTM_hh_home_TAZ.csv"))
    ld = pd.read_csv(os.path.join(IN_DIR, "LDPTM_hh_home_TAZ.csv"))
    ld_ae = pd.read_csv(os.path.join(IN_DIR, "LDPTM_AccEgr_hh_home_TAZ.csv"))
    etm = pd.read_csv(os.path.join(IN_DIR, "ETM_hh_home_TAZ.csv"))
    share = pd.read_csv(SHARE_CSV)

    ca_share = (
        share.groupby("year", as_index=False)
        .agg(EVhh_CA=("EVhh_TAZ", "sum"), hh_CA=("hh_TAZ", "sum"))
    )
    ca_share["EVhh_share"] = ca_share["EVhh_CA"] / ca_share["hh_CA"]

    def _count_by_zone(df: pd.DataFrame) -> pd.DataFrame:
        return df.groupby("HomeZone", as_index=False).size().rename(columns={"size": "n_hh"})

    sd_count = _calc_new_evhh(
        _expand_years(_count_by_zone(sd)),
        share[["year", "TAZ", "EVhh_share"]],
        join_cols=[("year", "year"), ("HomeZone", "TAZ")] if False else None,
    ) if False else None

    # Use pd.merge with left_on/right_on for clarity
    def _make_count(df, share_df, by_taz=True):
        c = _expand_years(_count_by_zone(df))
        if by_taz:
            out = pd.merge(c, share_df, left_on=["year", "HomeZone"], right_on=["year", "TAZ"])
        else:
            out = pd.merge(c, share_df, on="year")
        out["n_EVhh"] = out["n_hh"] * out["EVhh_share"]
        out = out.sort_values(["HomeZone", "year"])
        out["n_EVhh_newf"] = out.groupby("HomeZone")["n_EVhh"].diff()
        out.loc[out["year"] == YEARS[0], "n_EVhh_newf"] = out.loc[
            out["year"] == YEARS[0], "n_EVhh"
        ]
        out["n_EVhh_new"] = np.floor(out["n_EVhh_newf"]).astype("Int64")
        return out

    sd_count = _make_count(sd, share[["year", "TAZ", "EVhh_share"]])
    ld_count = _make_count(ld, share[["year", "TAZ", "EVhh_share"]])
    ld_ae_count = _make_count(ld_ae, share[["year", "TAZ", "EVhh_share"]])
    etm_count = _make_count(etm, ca_share[["year", "EVhh_share"]], by_taz=False)

    # Long distance
    evhh_ld = sample_evhh(ld, ld_count)
    evhh_ld.to_csv(os.path.join(IN_DIR, "EVhh_new_LDPTM_sample42.csv"), index=False)

    # Short distance
    evhh_sd = sample_evhh(sd, sd_count)
    evhh_sd.to_csv(os.path.join(IN_DIR, "EVhh_new_SDPTM_sample42.csv"), index=False)

    # Long distance access/egress
    evhh_ld_ae = sample_evhh(ld_ae, ld_ae_count)
    evhh_ld_ae.to_csv(os.path.join(IN_DIR, "EVhh_new_LDPTM_AccEgr_sample42.csv"), index=False)

    # ETM
    evhh_etm = sample_evhh(etm, etm_count)
    evhh_etm.to_csv(os.path.join(IN_DIR, "EVhh_new_ETM_sample42.csv"), index=False)


if __name__ == "__main__":
    main()
